dpnp/backend/extensions/statistics/histogram_common.cpp

```diff
@@ -0,0 +1,240 @@
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include <algorithm>
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "dpctl4pybind11.hpp"
+#include "utils/memory_overlap.hpp"
+#include "utils/output_validation.hpp"
+#include "utils/type_dispatch.hpp"
+
+#include <pybind11/pybind11.h>
+
+#include "histogram_common.hpp"
+
+namespace dpctl_td_ns = dpctl::tensor::type_dispatch;
+using dpctl::tensor::usm_ndarray;
+using dpctl_td_ns::typenum_t;
+
+namespace statistics
+{
+using common::CeilDiv;
+
+namespace histogram
+{
+
+void validate(const usm_ndarray &sample,
+              const usm_ndarray &bins,
+              std::optional<const dpctl::tensor::usm_ndarray> &weights,
+              const usm_ndarray &histogram)
+{
+    auto exec_q = sample.get_queue();
+    using array_ptr = const usm_ndarray *;
+
+    std::vector<array_ptr> arrays{&sample, &bins, &histogram};
+    std::unordered_map<array_ptr, std::string> names = {
+        {arrays[0], "sample"}, {arrays[1], "bins"}, {arrays[2], "histogram"}};
+
+    array_ptr weights_ptr = nullptr;
+
+    if (weights.has_value()) {
+        weights_ptr = &weights.value();
+        arrays.push_back(weights_ptr);
+        names.insert({weights_ptr, "weights"});
+    }
+
+    auto get_name = [&](const array_ptr &arr) {
+        auto name_it = names.find(arr);
+        assert(name_it != names.end());
+
+        return "'" + name_it->second + "'";
+    };
+
+    dpctl::tensor::validation::CheckWritable::throw_if_not_writable(histogram);
+
+    auto unequal_queue =
+        std::find_if(arrays.cbegin(), arrays.cend(), [&](const array_ptr &arr) {
+            return arr->get_queue() != exec_q;
+        });
+
+    if (unequal_queue != arrays.cend()) {
+        throw py::value_error(
+            get_name(*unequal_queue) +
+            " parameter has incompatible queue with parameter " +
+            get_name(&sample));
+    }
+
+    auto non_contig_array =
+        std::find_if(arrays.cbegin(), arrays.cend(), [&](const array_ptr &arr) {
+            return !arr->is_c_contiguous();
+        });
+
+    if (non_contig_array != arrays.cend()) {
+        throw py::value_error(get_name(*non_contig_array) +
+                              " parameter is not c-contiguos");
+    }
+
+    auto check_overlaping = [&](const array_ptr &first,
+                                const array_ptr &second) {
+        const auto &overlap = dpctl::tensor::overlap::MemoryOverlap();
+
+        if (overlap(sample, histogram)) {
```

Comment:
I guess it should be
```suggestion
        if (overlap(first, second)) {
```

Reply:
Thanks for noticing this!
Fixed